File: Leetcode/21. Merge Two Sorted Lists/Solution.java
/* https://leetcode.com/problems/merge-two-sorted-lists/

Merge two sorted linked lists and return it as a new list. 
The new list should be made by splicing together the nodes of the first two lists.
*/

class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {   
        ListNode merged = new ListNode(0);
        ListNode mergedptr = merged;
        
        while(l1 != null && l2 != null) {
            if(l1.val < l2.val) {
                mergedptr.next = l1;
                l1 = l1.next;
                mergedptr = mergedptr.next;
            }
            else{
                mergedptr.next = l2;
                l2 = l2.next;
                mergedptr = mergedptr.next;
            }
        }
        if(l1 != null) mergedptr.next = l1; //get the rest
        if(l2 != null) mergedptr.next = l2; //get the rest

        return merged.next;
    }
}

class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        if(l1 == null)
            return l2;
        if(l2 == null)
            return l1;
        
        if(l1.val < l2.val) {
            l1.next = mergeTwoLists(l1.next, l2);
            return l1;
        }
        else {
            l2.next = mergeTwoLists(l1, l2.next);
            return l2;
        }
    }
}

File: Leetcode/21. Merge Two Sorted Lists/Solution.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} list1
 * @param {ListNode} list2
 * @return {ListNode}
 */
 var mergeTwoLists = function(list1, list2) {
  let newHead = new ListNode();
  let mergedList = newHead;

  while(list1 !== null && list2 !== null) {
      if(list1.val < list2.val) {
          mergedList.next = list1;

          list1 = list1.next;
          mergedList = mergedList.next;
      }
      else {
          mergedList.next = list2;

          list2 = list2.next;
          mergedList = mergedList.next;
      }
  }

  if(list1 !== null) {
      mergedList.next = list1;
  }
  if(list2 !== null) {
      mergedList.next = list2;
  }
  return newHead.next;
};